### PR TITLE
Allow testing to update ValidatorSet

### PR DESCRIPTION
**Pull request by AdityaSripal**
## Description

closes: #1012 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: testing/chain.go [hunk 1]
```diff
@@ -261,13 +262,18 @@ func (chain *TestChain) QueryConsensusStateProof(clientID string) ([]byte, clien
 
 // NextBlock sets the last header to the current header and increments the current header to be
 // at the next block height. It does not update the time as that is handled by the Coordinator.
-//
-// CONTRACT: this function must only be called after app.Commit() occurs
 func (chain *TestChain) NextBlock() {
+	res := chain.App.EndBlock(abci.RequestEndBlock{Height: chain.CurrentHeader.Height})
+
+	chain.App.Commit()
```
Comment (AdityaSripal) [inline testing/chain.go:268]: EndBlock and Commit are now called within `NextBlock`. This allows us to correctly track validator changes as seen below

File: testing/simapp/test_helpers.go [hunk 2]
```diff
@@ -234,6 +234,8 @@ func CheckBalance(t *testing.T, app *SimApp, addr sdk.AccAddress, balances sdk.C
 
 // SignAndDeliver signs and delivers a transaction. No simulation occurs as the
 // ibc testing package causes checkState and deliverState to diverge in block time.
+//
+// CONTRACT: BeginBlock must be called before this function.
 func SignAndDeliver(
```
Comment (AdityaSripal) [inline testing/simapp/test_helpers.go:239]: SignAndDeliver will no longer simulate a block internally. It just calls DeliverTx and assumes BeginBlock has been called earlier

File: testing/utils.go [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	abci "github.com/tendermint/tendermint/abci/types"
+	tmtypes "github.com/tendermint/tendermint/types"
+)
+
+func ApplyValSetChanges(t *testing.T, valSet *tmtypes.ValidatorSet, valUpdates []abci.ValidatorUpdate) *tmtypes.ValidatorSet {
```
Comment (AdityaSripal) [inline testing/utils.go:11]: helper function to apply valset changes

Comment (colin-axner) [reply testing/utils.go:11]: Could you add godoc?

File: testing/utils.go [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+	updates, err := tmtypes.PB2TM.ValidatorUpdates(valUpdates)
+	require.NoError(t, err)
+
+	newVals := valSet.Copy()
```
Comment (colin-axner) [inline testing/utils.go:15]: might be good to add a comment for why this is necessary

File: testing/chain_test.go [hunk 5]
```diff
@@ -4,44 +4,32 @@ import (
...
+	coord.CommitBlock(chainA)
 
-	// smaller address
-	validator1.Address = []byte{1}
-	validator2.Address = []byte{2}
-	validator2.VotingPower = 1
+	path.EndpointB.UpdateClient()
 
-	expected = []tmtypes.PrivValidator{privVal1, privVal2}
+	path.EndpointB.UpdateClient()
```
Comment (colin-axner) [inline testing/chain_test.go:33]: should you be checking that the validator set was correctly updated?

File: testing/chain.go [hunk 6]
```diff
@@ -457,7 +464,12 @@ func (chain *TestChain) CreateTMClientHeader(chainID string, blockHeight int64,
 	blockID := MakeBlockID(hhash, 3, tmhash.Sum([]byte("part_set")))
 	voteSet := tmtypes.NewVoteSet(chainID, blockHeight, 1, tmproto.PrecommitType, tmValSet)
 
-	commit, err := tmtypes.MakeCommit(blockID, blockHeight, 1, voteSet, signers, timestamp)
+	var signerArr []tmtypes.PrivValidator
+	for _, v := range tmValSet.Validators {
```
Comment (colin-axner) [inline testing/chain.go:468]: is it going to be problematic that you iterate over a map which may not give correct ordering of the validators?

File: testing/chain.go [hunk 7]
```diff
@@ -311,7 +317,7 @@ func (chain *TestChain) SendMsgs(msgs ...sdk.Msg) (*sdk.Result, error) {
 		return nil, err
 	}
 
-	// SignAndDeliver calls app.Commit()
+	// NextBlock app.Commit()
```
Comment (colin-axner) [inline testing/chain.go:320]: remove comment?

File: testing/chain.go (hunk 1 above)
Comment (colin-axner) [reply testing/chain.go:268]: Can you add this information to the godoc?

File: testing/chain.go [hunk 8]
```diff
@@ -261,13 +262,18 @@ func (chain *TestChain) QueryConsensusStateProof(clientID string) ([]byte, clien
...
+
+	chain.App.Commit()
+
 	// set the last header to the current header
 	// use nil trusted fields
 	chain.LastHeader = chain.CurrentTMClientHeader()
 
+	chain.Vals = chain.NextVals
+	chain.NextVals = ApplyValSetChanges(chain.T, chain.Vals, res.ValidatorUpdates)
+
```
Comment (colin-axner) [inline testing/chain.go:276]: can you add documentation why the validator set changes from the last committed block won't take place until after committing the current block?

File: testing/chain.go [hunk 9]
```diff
@@ -60,8 +59,15 @@ type TestChain struct {
 	TxConfig      client.TxConfig
 	Codec         codec.BinaryCodec
 
-	Vals    *tmtypes.ValidatorSet
-	Signers []tmtypes.PrivValidator
+	Vals     *tmtypes.ValidatorSet
+	NextVals *tmtypes.ValidatorSet
+
+	// Signers is a map from validator address to the PrivValidator
+	// The map is converted into an array that is the same order as the validators right before signing commit
```
Comment (colin-axner) [inline testing/chain.go:66]: Where do you ensure the ordering is the same? Go maps don't use any sorted ordering. You'd need to sort the array

File: testing/chain_test.go [hunk 10]
```diff
@@ -4,44 +4,32 @@ import (
...
-	validator2 := tmtypes.NewValidator(pubKey2, 2)
+	amount, ok := sdk.NewIntFromString("10000000000000000000")
+	require.True(t, ok)
 
-	expected := []tmtypes.PrivValidator{privVal2, privVal1}
+	val := chainA.App.GetStakingKeeper().GetValidators(chainA.GetContext(), 4)
 
-	actual := ibctesting.CreateSortedSignerArray(privVal1, privVal2, validator1, validator2)
-	require.Equal(t, expected, actual)
+	chainA.App.GetStakingKeeper().Delegate(chainA.GetContext(), chainA.SenderAccounts[1].SenderAccount.GetAddress(),
```
Comment (colin-axner) [inline testing/chain_test.go:26]: maybe you should be changing the power of multiple validators?

File: testing/chain.go [hunk 11]
```diff
@@ -457,7 +464,12 @@ func (chain *TestChain) CreateTMClientHeader(chainID string, blockHeight int64,
 	blockID := MakeBlockID(hhash, 3, tmhash.Sum([]byte("part_set")))
 	voteSet := tmtypes.NewVoteSet(chainID, blockHeight, 1, tmproto.PrecommitType, tmValSet)
 
-	commit, err := tmtypes.MakeCommit(blockID, blockHeight, 1, voteSet, signers, timestamp)
+	var signerArr []tmtypes.PrivValidator
+	for _, v := range tmValSet.Validators {
+		signerArr = append(signerArr, signers[v.Address.String()])
+	}
+
+	commit, err := tmtypes.MakeCommit(blockID, blockHeight, 1, voteSet, signerArr, timestamp)
```
Comment (colin-axner) [inline testing/chain.go:472]: Does MakeCommit sort the signerArr?

File: testing/chain.go (hunk 6 above)
Comment (AdityaSripal) [reply testing/chain.go:468]: No the validators is not a map. It is an ordered array.

We then have an unordered map of signers which we convert into an array of signers that is ordered the same way as validators array

File: testing/chain.go (hunk 9 above)
Comment (AdityaSripal) [reply testing/chain.go:66]: We convert the unordered map into ordered array here: https://github.com/cosmos/ibc-go/blob/aditya%2Fupdate-val-set/testing/chain.go#L467

File: testing/chain.go (hunk 11 above)
Comment (AdityaSripal) [reply testing/chain.go:472]: No, the signer arr is passed in with the order of the voteset. see above for loop

File: testing/chain.go (hunk 6 above)
Comment (colin-axner) [reply testing/chain.go:468]: Can you add a comment explaining this?

File: testing/chain.go (hunk 7 above)
Comment (AdityaSripal) [reply testing/chain.go]: I kept it. Just fixed the comment

**Comment by AdityaSripal** [comment]
@sainoe tested valset changes in tests using the changes in this PR and everything works 🎉

**Review by colin-axner — APPROVED** [review]
Wahoo!! Thanks for fixing this up! :)

**Review by damiannolan — APPROVED** [review]

**Review by seantking — APPROVED** [review]

**Pull request merged**